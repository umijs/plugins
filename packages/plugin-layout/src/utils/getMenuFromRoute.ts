/**
 * 通过路由配置自动生成菜单配置
 */
import { flatten } from 'lodash';
import {
  IBestAFSRoute,
  MenuItem,
  IRouteMenuConfig,
} from '../types/interface.d';

/**
 * Conversion router to menu.
 * - menu 中的全部参数，支持平铺式写法, menu 中的优先级更高
 * - indexRoute 已经被前置插件处理打平处理掉了
 * - menu 国际化
 */
function formatter(
  baseRoutes: IBestAFSRoute[] = [],
  prefix: string = '',
  base: string = '/',
): MenuItem[] {
  const menus = flatten(
    baseRoutes
      .filter(
        item =>
          item &&
          !item.unaccessible && // 是否没有权限查看
          item.menu !== false && // 显示指定在 menu 中隐藏该项
          (item.name || // 兼容老版本 route 配置
          item.flatMenu || // 是否打平 menu
          (item.menu && (item.menu.name || item.menu.flatMenu)) || // 正确配置了 menu
            (item.indexRoute &&
              item.indexRoute.menu &&
              (item.indexRoute.menu.name || item.indexRoute.menu.flatMenu))), // 有 indexRoute 且正确配置了中的 menu
      )
      .map(route => {
        const { menu = {}, indexRoute, path = '', routes } = route;
        const {
          name = route.name,
          icon = route.icon,
          hideChildren = route.hideChildren,
          flatMenu = route.flatMenu,
        } = menu as IRouteMenuConfig; // 兼容平铺式写法

        // 拼接 path
        const absolutePath =
          path.startsWith('/') || path.startsWith('http')
            ? path
            : `${base}${base === '/' ? '' : '/'}${path}`;

        // 拼接 childrenRoutes, 处理存在 indexRoute 时的逻辑
        const childrenRoutes = indexRoute
          ? [
              {
                path,
                menu,
                ...indexRoute,
              },
            ].concat(routes || [])
          : routes;

        // 拼接返回的 menu 数据
        const result = {
          ...route,
          name,
          path: path.startsWith('http')
            ? absolutePath
            : `${prefix}${absolutePath}`,
        } as MenuItem;

        if (icon) {
          result.icon = icon;
        }

        if (childrenRoutes && childrenRoutes.length) {
          /** 在菜单中隐藏子项 */
          if (hideChildren) {
            delete result.children;
            return result;
          }

          const children = formatter(childrenRoutes, prefix, absolutePath);

          /** 在菜单中只隐藏此项，子项往上提，仍旧展示 */
          if (flatMenu) {
            return children;
          }

          result.children = children;
          // delete result.path;
        }
        return result;
      }),
  );
  return menus;
}

export default formatter;
