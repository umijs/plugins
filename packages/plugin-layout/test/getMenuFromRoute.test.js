import getMenuFromRoute from '../src/utils/getMenuFromRoute';
import { normalRoute, normalMenu } from './routes/normal';
import { simpleRoute, simpleMenu } from './routes/simple';
import { prefixRoute, prefixMenu } from './routes/prefix';

describe('getMenuFromRoute', () => {
  /**
   * 推荐写法
   * 测试了以下场景：
   *  1. 子路由
   *  2. 兼容老的 name / icon 写法
   *  3. 相对路径 / 绝对路径
   *  4. indexRoute
   *  5. flatMenu 打平菜单
   *  6. 隐藏菜单
   */
  test('normal', () => {
    const menu = getMenuFromRoute(normalRoute);
    expect(menu).toMatchSnapshot();
  });

  /**
   * 简易写法
   * 测试了以下场景:
   *  1. 平铺式写法，兼容老的配置写法
   *  2. hideChildren 隐藏子菜单
   */
  test('simple', () => {
    const menu = getMenuFromRoute(simpleRoute);
    expect(menu).toMatchSnapshot();
  });

  /**
   * 前缀
   * 测试了以下场景：
   *  1. 正确生成 OneX 需要的 mock 菜单数据
   */
  test('prefix', () => {
    const menu = getMenuFromRoute(prefixRoute, '#/$name');
    expect(menu).toMatchSnapshot();
  });
});
