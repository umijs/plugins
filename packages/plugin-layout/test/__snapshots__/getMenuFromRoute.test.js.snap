// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getMenuFromRoute normal 1`] = `
Array [
  Object {
    "component": "IndexPage",
    "layout": false,
    "menu": Object {
      "name": "首页",
    },
    "name": "首页",
    "path": "/",
  },
  Object {
    "component": "IndexPage",
    "icon": "testicon",
    "menu": Object {
      "icon": "testicon",
      "name": "欢迎",
    },
    "name": "欢迎",
    "path": "/welcome",
  },
  Object {
    "children": Array [
      Object {
        "component": "IndexPage",
        "layout": Object {
          "hideMenu": true,
        },
        "menu": Object {
          "name": "示例 1",
        },
        "name": "示例 1",
        "path": "/test/1",
      },
      Object {
        "component": "IndexPage",
        "icon": "testicon",
        "layout": Object {
          "hideNav": true,
        },
        "menu": Object {
          "icon": "testicon",
          "name": "示例 2",
        },
        "name": "示例 2",
        "path": "/test233",
      },
    ],
    "menu": Object {
      "name": "菜单示例",
    },
    "name": "菜单示例",
    "path": "/test",
    "routes": Array [
      Object {
        "component": "IndexPage",
        "layout": Object {
          "hideMenu": true,
        },
        "menu": Object {
          "name": "示例 1",
        },
        "path": "1",
      },
      Object {
        "component": "IndexPage",
        "layout": Object {
          "hideNav": true,
        },
        "menu": Object {
          "icon": "testicon",
          "name": "示例 2",
        },
        "path": "/test233",
      },
      Object {
        "component": "IndexPage",
        "path": "/test_hide",
      },
    ],
  },
  Object {
    "menu": Object {
      "name": "关于 Bigfish",
    },
    "name": "关于 Bigfish",
    "path": "https://bigfish.antfin-inc.com/",
  },
]
`;

exports[`getMenuFromRoute prefix 1`] = `
Array [
  Object {
    "component": "IndexPage",
    "menu": Object {},
    "name": "首页",
    "path": "#/$name/",
  },
  Object {
    "component": "IndexPage",
    "icon": "testicon",
    "layout": Object {
      "hideMenu": false,
    },
    "name": "欢迎",
    "path": "#/$name/welcome",
  },
  Object {
    "children": Array [
      Object {
        "component": "IndexPage",
        "name": "示例 1",
        "path": "#/$name/test/1",
      },
      Object {
        "component": "IndexPage",
        "icon": "testicon",
        "layout": Object {
          "hideNav": false,
        },
        "name": "示例 2",
        "path": "#/$name/test233",
      },
    ],
    "layout": Object {
      "hideNav": true,
    },
    "name": "菜单示例",
    "path": "#/$name/test",
    "routes": Array [
      Object {
        "component": "IndexPage",
        "name": "示例 1",
        "path": "1",
      },
      Object {
        "component": "IndexPage",
        "icon": "testicon",
        "layout": Object {
          "hideNav": false,
        },
        "name": "示例 2",
        "path": "/test233",
      },
      Object {
        "component": "IndexPage",
        "path": "/test_hide",
      },
    ],
  },
  Object {
    "name": "关于 Bigfish",
    "path": "https://bigfish.antfin-inc.com/",
  },
]
`;

exports[`getMenuFromRoute simple 1`] = `
Array [
  Object {
    "component": "IndexPage",
    "menu": Object {},
    "name": "首页",
    "path": "/",
  },
  Object {
    "component": "IndexPage",
    "icon": "testicon",
    "layout": Object {
      "hideMenu": false,
    },
    "name": "欢迎",
    "path": "/welcome",
  },
  Object {
    "children": Array [
      Object {
        "component": "IndexPage",
        "name": "示例 1",
        "path": "/test/1",
      },
      Object {
        "component": "IndexPage",
        "icon": "testicon",
        "layout": Object {
          "hideNav": false,
        },
        "name": "示例 2",
        "path": "/test233",
      },
    ],
    "layout": Object {
      "hideNav": true,
    },
    "name": "菜单示例",
    "path": "/test",
    "routes": Array [
      Object {
        "component": "IndexPage",
        "name": "示例 1",
        "path": "1",
      },
      Object {
        "component": "IndexPage",
        "icon": "testicon",
        "layout": Object {
          "hideNav": false,
        },
        "name": "示例 2",
        "path": "/test233",
      },
      Object {
        "component": "IndexPage",
        "path": "/test_hide",
      },
    ],
  },
  Object {
    "name": "关于 Bigfish",
    "path": "https://bigfish.antfin-inc.com/",
  },
]
`;
